#164139683 Add password reset feature

youngestdj opened the pull request:
#### What does this PR do?
This feature adds a feature for a user to reset password
#### Description of task to be completed?
* Add `passwordResetToken` and `passwordResetTokenExpires` columns to `User` table
* Add endpoint to send reset email
* Add endpoint to verify password reset link
* Add endpoint to change user password
* Write tests
* Add authentication class
#### How should this be manually tested?
* Run `npm run db-migrate` to run migrations
* Run `npm start` to start the application
* Create a user by sending a post request to `/api/v1/user`
* Verify the user by clicking the link in your email
* Send a post request with your email to `/api/v1/reset` with your email as post body `email: 'email@domain.com'`
* Click the link sent to your email
* Copy the token in the request response
* Send token along with your new password to `api/v1/reset/3`
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/164139683

youngestdj committed:
feat(endpoint): add password reset endpoint

- add columns to user table
- add endpoint to send reset email
- add endpoint to verify password reset link
- add endpoint to change user password
- write tests
- add authentication class

[Delivers #164139683]

codinger41 commented:
Please Fix your merge conflicts @youngestdj

olajide1234 commented on controllers/ChangePassword.js:
```diff
@@ -0,0 +1,50 @@
+
+    try {
+      // delete password token
+      await User.update(
```
Hi. Please consider using `User.destroy()` to delete the passwordReserToken.
https://lorenstewart.me/2016/10/03/sequelize-crud-101/ 
http://docs.sequelizejs.com/class/lib/model.js~Model.html#instance-method-destroy

youngestdj replied:
`User.destroy` will delete the entire row. I'm not trying to delete the row on that line, only trying to remove the password token

olajide1234 commented on controllers/PasswordReset.js:
```diff
@@ -0,0 +1,83 @@
+   * @param {string} email user email
+   * @returns {Boolean} true or false
+   */
+  static async userExists(email) {
```
Consider extracting this to a helper function, so it can be available for others to use.

olajide1234 commented on controllers/PasswordReset.js:
```diff
@@ -0,0 +1,83 @@
+   * @param {string} email user email
+   * @returns {promise} promise object
+   */
+  static async getName(email) {
```
Consider extracting this to a helper function, so it can be available for others to use.

olajide1234 commented on routes/v1.js:
```diff
@@ -16,4 +25,24 @@ apiRoutes.get(
 );
 
+apiRoutes.post(
+  '/reset',
```
Consider a more descriptive route name

olajide1234 commented on test/resetpassword.spec.js:
```diff
@@ -0,0 +1,167 @@
+chai.use(chaiHttp);
+const { expect } = chai;
+
+const validUser = {
```
Consider extracting the test data to a `test\helpers\dummyData.js` file. Also, `verified: true` may not pass through the sign-up route again. Let's sync in person.

olajide1234 reviewed (commented):
Great job Samuel. Please look at these few comments. Thanks.

youngestdj committed:
feat(endpoint): add password reset feature

- add columns to user table
- add endpoint to send reset email
- add endpoint to verify password reset link and new password
- write tests

[Delivers #164139683]

9jaswag commented on .nyc_output/76c64d62-a9ed-41c2-80f1-279e09c5da1d.json:
```diff
@@ -0,0 +1 @@
+{}
```
@youngestdj  shouldn't this be git ignored?

youngestdj replied:
I have added it to `.gitignore`

9jaswag commented on routes/v1.js:
```diff
@@ -2,15 +2,19 @@ import express from 'express';
+  returnValidationErrors,
+  validatePassword
 } from '../middleware/validation';
+import PasswordReset from '../controllers/PasswordReset';
```
This should be in the user controller

youngestdj replied:
I have moved the password reset methods to the user controller

9jaswag commented:
these should be the last set of changes. This PR should be merge ready afterward.

youngestdj committed:
feat(endpoint): add password reset feature

- add columns to user table
- add endpoint to send reset email
- add endpoint to verify password reset link and new password
- write tests

[Delivers #164139683]

youngestdj commented:
> these should be the last set of changes. This PR should be merge ready afterward.

9jaswag commented on controllers/user.js:
```diff
@@ -131,4 +136,102 @@ export default class UserController {
+    try {
+      // save password reset token to db
+      await User.update(
+        { passwordResetToken },
```
@youngestdj I mentioned that this update query should be done at once and not in two different places

youngestdj replied:
I have used a single update query

9jaswag replied:
@youngestdj I can't see that here. There's another update on line 157. That's what I'm referring to

9jaswag commented on routes/v1.js:
```diff
@@ -76,4 +78,19 @@ apiRoutes.post(
 );
 
+apiRoutes.post(
+  '/resetpassword',
```
@youngestdj this route can be better named, since it is not resetting the password but requesting for a password reset. This route name is better for the second route since it actually resets the password.
Ensure this change reflects in the controller method name too.

youngestdj replied:
I have renamed the route and controller names

9jaswag commented on controllers/user.js:
```diff
@@ -75,7 +80,7 @@ export default class UserController {
       user
     } = req;
-    const passwordMatch = compareSync(password, user.password);
+    const passwordMatch = bcrypt.compareSync(password, user.password);
```
@youngestdj are you sure this change works?

youngestdj replied:
Yes it does. I have manually tested it and tests that utilize that line pass

9jaswag commented on controllers/user.js:
```diff
@@ -131,4 +136,93 @@ export default class UserController {
+   * @param {object} res http response object
+   * @returns {object} response
+   */
+  static async requestpasswordreset(req, res) {
```
@youngestdj this should be `requestPasswordReset`

youngestdj replied:
I have implemented changes

9jaswag commented on controllers/user.js:
```diff
@@ -131,4 +136,93 @@ export default class UserController {
+   * @param {object} res http response object
+   * @returns {object} response
+   */
+  static async resetpassword(req, res) {
```
@youngestdj this should be `resetPassword`. Let's stick to the naming convention

youngestdj replied:
Fixed

9jaswag commented on doc.json:
```diff
@@ -343,10 +437,10 @@
-    "url": "http://swagger.io"
   }
-}
+}
```
@youngestdj Fix this EOL issue

youngestdj replied:
Fixed

9jaswag commented on doc.json:
```diff
@@ -231,6 +231,84 @@
+        }
+      }
+    },
+    "/verifypasswordkey/{key}": {
```
@youngestdj you did not update the swagger docs with the right URLs

youngestdj replied:
Fixed
